manual/sphinx/developer_docs/performance_profiling.rst

```diff
@@ -0,0 +1,107 @@
+.. Use bash as the default language for syntax highlighting in this file
+.. highlight:: console
+
+.. _sec-performanceprofiling:
+
+Performance profiling
+=====================
+
+Analyzing code behaviour is vital for getting the best performance from BOUT++.
+This is done by profiling the code, that is, building and running the code 
+using tools that report the amount of time each processor spends in functions,
+on communications, etc.
+
+This section describes how to compile and run BOUT++ using the 
+`Scorep <http://www.vi-hps.org/projects/score-p/>`_/`Scalasca <http://www.scalasca.org/>`_
+tool chain.
+
+Scorep/Scalasca profiling
+-------------------------
+
+Instrumentation
+~~~~~~~~~~~~~~~
+
+Scorep automatically reports the time spend in MPI communications and OpenMP
+loops. However, to obtain information on the time spent in specific functions,
+it is necessary to instrument the source code. The macros to do this are 
+provided in ``scorepwrapper.hxx``.
+
+To include a function in Scorep's timing, include the scorep wrapper in the 
+source code
+
+.. code-block:: c++
+
+    #include <bout/scorepwrapper.hxx>
+
+and then write the macro ``SCOREP0()`` at the top of the function, e.g.
+
+.. code-block:: c++
+
+    int Field::getNx() const{
+      SCOREP0();
+      return getMesh()->LocalNx;
+    };
+
+**Caution** Instrumenting a function makes it execute more slowly. This can
+result in misleading profiling information, particularly if 
+fast-but-frequently-called functions are instrumented. Try to instrument 
+significant functions only.
+
+This notwithstanding, it is reasonable to expect sensibly-instrumented code to
+run about 50% to 100% slower than the uninstrumented code.
```

Comment:
I'm not sure about this -- I think if we have this much slow down then it's going to be difficult to really interpret things, I usually find a much smaller reduction in performance when the profiling is limited to things outside of loops over the mesh.

Reply:
Ah okay - these figures are for profiling operators on fields.  What would you say is a reasonable overhead for less intensive profiling?

Reply:
Thanks - updated.